src/Orleans/Runtime/GrainReference.cs

```diff
@@ -531,14 +551,21 @@ protected internal static object DeserializeGrainReference(Type t, IDeserializat
             }
             // store as null, serialize as empty.
             var genericArg = reader.ReadString();
-            if (String.IsNullOrEmpty(genericArg))
+            if (string.IsNullOrEmpty(genericArg))
                 genericArg = null;
 
+            GrainReference result;
             if (expectObserverId)
             {
-                return NewObserverGrainReference(id, observerId);
+                result = NewObserverGrainReference(id, observerId, null);
+            }
+            else
+            {
+                result = FromGrainId(id, null, genericArg, silo);
             }
-            return FromGrainId(id, genericArg, silo);
+
+            context.GrainFactory?.BindGrainReference(result);
```

Comment:
could there be gotchas with this? I presume not, but maybe in a multi-cluster scenario or something of that sort?